test-grapesy/Test/Sanity/Disconnect.hs

```diff
@@ -0,0 +1,288 @@
+
+    -- We should also see only 50 client disconnects for the first handler and
+    -- none for the second
+    clientDisconnects1 <- readIORef disconnectCounter1
+    clientDisconnects2 <- readIORef disconnectCounter2
+    assertBool "" (clientDisconnects1 == 50 && clientDisconnects2 == 0)
+
+-- | Client makes many concurrent calls, server disconnects
+test_serverDisconnect :: Assertion
+test_serverDisconnect = withTemporaryFile $ \ipcFile -> do
+    -- We use a temporary file as a very rudimentary means of inter-process
+    -- communication so the server (which runs in a separate process) can make
+    -- the client aware of the port it is assigned by the OS. This also helps us
+    -- make sure the server binds to the same port when it comes back up for
+    -- reconnect purposes.
+    let ipcWrite :: String -> IO ()
+        ipcWrite msg = do
+          writeFile ipcFile ""
+          writeFile ipcFile msg
+
+        ipcRead :: IO String
+        ipcRead = readFile ipcFile
+
+        ipcWaitRead :: IO String
+        ipcWaitRead = do
+          ipcRead >>= \case
+            "" -> do
+              threadDelay 10000 >> ipcWaitRead
+            msg -> do
+              return msg
+
+    -- Create the server
+    server <-
+      Server.mkGrpcServer def [
+          Server.someRpcHandler $
+              Server.mkRpcHandler @Trivial $ echoHandler Nothing
+        ]
+
+    let -- Starts the server in a new process. Gives back an action that kills
+        -- the server process.
+        startServer :: IO (IO ())
+        startServer = do
+          serverPid <-
+            forkProcess $
+              rawTestServer (readMaybe <$> ipcRead) (ipcWrite . show) server
+          return $ c_kill (fromIntegral serverPid) sigKILL
+
+    -- Start server, get the port
+    killServer    <- startServer
+    serverPort    <- read <$> ipcWaitRead
+    signalRestart <- newEmptyMVar
+    let serverAddress =
+          Client.ServerInsecure Client.Address {
+              addressHost      = "127.0.0.1"
+            , addressPort      = serverPort
+            , addressAuthority = Nothing
+            }
+
+        reconnectPolicy :: Client.ReconnectPolicy
+        reconnectPolicy = go 0
+          where
+            go :: Int -> Client.ReconnectPolicy
+            go n
+              | n == 5
+              = Client.ReconnectAfter $ do
+                  killRestarted <- startServer
+                  putMVar signalRestart killRestarted
+                  return $ Client.exponentialBackoff threadDelay 1 (1, 1) 100
+              | otherwise
+              = Client.ReconnectAfter $ do
+                  threadDelay 10000
+                  return $ go (n + 1)
+
+        connParams :: Client.ConnParams
+        connParams = def { Client.connReconnectPolicy = reconnectPolicy }
+
+    Client.withConnection connParams serverAddress $ \conn -> do
+      -- Make 50 concurrent calls. 49 of them sending infinite messages. One
```

Comment:
Some duplication here with the first test. Not a huge deal, but I actually think that abstracting this out would not just avoid the duplication but also clarify the test.

Reply:
Will abstract 🫡

Reply:
Test steps are now described by a tiny DSL and interpreted, no more predicates or pure exceptions. Feels much cleaner.